tpch/tpcbench.py

```diff
@@ -0,0 +1,97 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+import argparse
+import ray
+from datafusion_ray import DatafusionRayContext
+from datetime import datetime
+import json
+import time
+
+def main(benchmark: str, data_path: str, query_path: str, concurrency: int):
+
+    # Register the tables
+    if benchmark == "tpch":
+        num_queries = 22
+        table_names = ["customer", "lineitem", "nation", "orders", "part", "partsupp", "region", "supplier"]
+    elif benchmark == "tpcds":
+        num_queries = 99
+        table_names = ["call_center", "catalog_page", "catalog_returns", "catalog_sales", "customer",
+           "customer_address", "customer_demographics", "date_dim", "time_dim", "household_demographics",
+           "income_band", "inventory", "item", "promotion", "reason", "ship_mode", "store", "store_returns",
+           "store_sales", "warehouse", "web_page", "web_returns", "web_sales", "web_site"]
+    else:
+        raise "invalid benchmark"
+
+    # Connect to a cluster
+    # use ray job submit
+    ray.init()
+
+    ctx = DatafusionRayContext(concurrency)
+
+    for table in table_names:
+        path = f"{data_path}/{table}.parquet"
+        print(f"Registering table {table} using path {path}")
+        ctx.register_parquet(table, path)
+
+    results = {
+        'engine': 'datafusion-python',
+        'benchmark': benchmark,
+        'data_path': data_path,
+        'query_path': query_path,
+        'concurrency': concurrency,
+    }
+
+    for query in range(1, num_queries + 1):
+        # read text file
+        path = f"{query_path}/q{query}.sql"
+        print(f"Reading query {query} using path {path}")
+        with open(path, "r") as f:
+            text = f.read()
+            # each file can contain multiple queries
+            queries = text.split(";")
+
+            start_time = time.time()
+            for sql in queries:
+                sql = sql.strip()
+                if len(sql) > 0:
+                    print(f"Executing: {sql}")
+                    rows = ctx.sql(sql)
```

Comment:
How does this time relate to the fact that the context only wait for the first partition ? see https://github.com/apache/datafusion-ray/blob/main/datafusion_ray/context.py#L234

is there always a single partition?

Reply:
Yes, the final query stage always has a single output partition. You can verify this by looking at the expected plans in the repo.